src/main/kotlin/org/rust/cargo/project/configurable/RsExternalLinterConfigurable.kt

```diff
@@ -11,33 +11,41 @@ import com.intellij.ui.EnumComboBoxModel
 import com.intellij.ui.dsl.builder.*
 import org.rust.RsBundle
 import org.rust.cargo.project.model.cargoProjects
+import org.rust.cargo.project.settings.RsExternalLinterProjectSettingsService
+import org.rust.cargo.project.settings.externalLinterSettings
 import org.rust.cargo.toolchain.ExternalLinter
 import org.rust.cargo.util.CargoCommandCompletionProvider
 import org.rust.cargo.util.RsCommandLineEditor
 import org.rust.openapiext.fullWidthCell
 
 class RsExternalLinterConfigurable(project: Project) : RsConfigurableBase(project, RsBundle.message("settings.rust.external.linters.name")) {
+    private val settings: RsExternalLinterProjectSettingsService = project.externalLinterSettings
+
+    private val additionalArguments: RsCommandLineEditor =
+        RsCommandLineEditor(project, CargoCommandCompletionProvider(project.cargoProjects, "check ") { null })
+
     override fun createPanel(): DialogPanel = panel {
         row(RsBundle.message("settings.rust.external.linters.tool.label")) {
             comboBox(EnumComboBoxModel(ExternalLinter::class.java))
                 .comment(RsBundle.message("settings.rust.external.linters.tool.comment"))
-                .bindItem(state::externalLinter.toNullableProperty())
+                .bindItem(settings.state::tool.toNullableProperty())
         }
 
         row(RsBundle.message("settings.rust.external.linters.additional.arguments.label")) {
-            fullWidthCell(RsCommandLineEditor(project, CargoCommandCompletionProvider(project.cargoProjects, "check ") { null }))
+            fullWidthCell(additionalArguments)
+                .resizableColumn()
                 .comment(RsBundle.message("settings.rust.external.linters.additional.arguments.comment"))
                 .bind(
                     componentGet = { it.text },
                     componentSet = { component, value -> component.text = value },
-                    prop = state::externalLinterArguments.toMutableProperty()
+                    prop = settings.state::additionalArguments.toMutableProperty()
                 )
         }
 
         row {
             checkBox(RsBundle.message("settings.rust.external.linters.on.the.fly.label"))
                 .comment(RsBundle.message("settings.rust.external.linters.on.the.fly.comment"))
-                .bindSelected(state::runExternalLinterOnTheFly)
+                .bindSelected(settings.state::runOnTheFly)
```

Comment:
Found the problem - here (and in all similar places) you modify the state directly, not via `modify` method. As a result, you don't invoke `notifySettingsChanged` and nobody notices the change. For example, external linter widget status doesn't change

In general, it shows the problem of the current design - there are two ways to modify settings, via `modify` and directly, and you are not forced to use one of them. If we keep both ways, it 100% would lead to such cases as here